File: android/app/src/main/java/com/rnappota/ota/OtaBundleManager.kt
package com.rnappota.ota

import android.content.Context
import android.content.SharedPreferences
import android.util.Log
import java.io.File

/**
 * OtaBundleManager
 * ────────────────
 * Singleton that reads / writes OTA bundle paths from SharedPreferences.
 * Called from MainApplication BEFORE the React runtime starts, so it must
 * be pure Kotlin — no RN dependencies.
 *
 * Keys:
 *   ota_active_bundle_path    — absolute path of the currently active OTA bundle (or null)
 *   ota_pending_bundle_path   — absolute path of a downloaded-but-not-yet-active bundle
 *   ota_previous_bundle_path  — absolute path of the bundle before the current one (rollback)
 *   ota_crash_count           — number of consecutive cold-start crashes
 */
object OtaBundleManager {

    private const val TAG = "OTA"
    private const val PREFS_NAME = "ota_prefs"
    private const val KEY_ACTIVE = "ota_active_bundle_path"
    private const val KEY_PENDING = "ota_pending_bundle_path"
    private const val KEY_PREVIOUS = "ota_previous_bundle_path"
    private const val KEY_CRASH_COUNT = "ota_crash_count"

    private lateinit var prefs: SharedPreferences

    fun init(context: Context) {
        prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    }

    // ─── Active bundle ────────────────────────────────────────────────────────

    /**
     * Returns the filesystem path of the active OTA bundle, or null if the
     * app should load the bundled APK asset (default behaviour).
     * Also validates the file exists; clears the path and returns null if not.
     */
    fun getActiveBundlePath(): String? {
        val path = prefs.getString(KEY_ACTIVE, null) ?: run {
            Log.d(TAG, "No active OTA bundle — loading from APK asset.")
            return null
        }
        if (!File(path).exists()) {
            Log.w(TAG, "Active bundle path set but file missing: $path — clearing, falling back to APK asset.")
            clearActiveBundlePath()
            return null
        }
        Log.d(TAG, "Active OTA bundle: $path")
        return path
    }

    fun setActiveBundlePath(path: String) {
        prefs.edit().putString(KEY_ACTIVE, path).apply()
    }

    fun clearActiveBundlePath() {
        prefs.edit().remove(KEY_ACTIVE).apply()
    }

    // ─── Pending bundle ───────────────────────────────────────────────────────

    fun getPendingBundlePath(): String? = prefs.getString(KEY_PENDING, null)

    fun setPendingBundlePath(path: String) {
        prefs.edit().putString(KEY_PENDING, path).apply()
    }

    fun clearPendingBundlePath() {
        prefs.edit().remove(KEY_PENDING).apply()
    }

    // ─── Previous bundle (rollback target) ────────────────────────────────────

    fun getPreviousBundlePath(): String? = prefs.getString(KEY_PREVIOUS, null)

    fun setPreviousBundlePath(path: String?) {
        if (path == null) {
            prefs.edit().remove(KEY_PREVIOUS).apply()
        } else {
            prefs.edit().putString(KEY_PREVIOUS, path).apply()
        }
    }

    // ─── Crash counter ────────────────────────────────────────────────────────

    fun getCrashCount(): Int = prefs.getInt(KEY_CRASH_COUNT, 0)

    fun incrementCrashCount() {
        val count = getCrashCount()
        val next = count + 1
        prefs.edit().putInt(KEY_CRASH_COUNT, next).apply()
        Log.d(TAG, "Crash count incremented: $count → $next")
    }

    fun resetCrashCount() {
        prefs.edit().putInt(KEY_CRASH_COUNT, 0).apply()
    }

    // ─── Rollback ─────────────────────────────────────────────────────────────

    /**
     * Atomically rolls back: sets active to previous, clears pending.
     * Called from the crash guard path when JS keeps crashing.
     */
    fun rollback() {
        val previous = getPreviousBundlePath()
        if (previous != null && File(previous).exists()) {
            Log.d(TAG, "Rollback — restoring previous bundle: $previous")
            setActiveBundlePath(previous)
        } else {
            Log.w(TAG, "Rollback — no valid previous bundle found, falling back to APK asset.")
            clearActiveBundlePath()           // fall back to APK asset
        }
        clearPendingBundlePath()
        prefs.edit().remove(KEY_PREVIOUS).apply()
    }

    // ─── OTA files directory ──────────────────────────────────────────────────

    /**
     * Returns (creating if needed) /data/data/<pkg>/files/ota/<label>/
     */
    fun getOrCreateOtaDir(filesDir: File, label: String): File {
        val dir = File(filesDir, "ota/$label")
        if (!dir.exists()) dir.mkdirs()
        return dir
    }
}

File: ios/RNAppOta/OtaBundleManager.swift
import Foundation

/**
 * OtaBundleManager (iOS)
 * ──────────────────────
 * Singleton that reads / writes OTA bundle paths from UserDefaults.
 * Called from AppDelegate.bundleURL() BEFORE any JS executes.
 */
@objc public class OtaBundleManager: NSObject {

    @objc public static let shared = OtaBundleManager()

    private let defaults = UserDefaults.standard

    private enum Key {
        static let active   = "ota_active_bundle_path"
        static let pending  = "ota_pending_bundle_path"
        static let previous = "ota_previous_bundle_path"
        static let crashCount = "ota_crash_count"
    }

    // ─── Active bundle ────────────────────────────────────────────────────────

    /// The URL of the active OTA bundle, or nil to use the shipped main.jsbundle.
    @objc public var activeBundleURL: URL? {
        guard let path = defaults.string(forKey: Key.active) else { return nil }
        let url = URL(fileURLWithPath: path)
        guard FileManager.default.fileExists(atPath: path) else {
            clearActiveBundlePath()
            return nil
        }
        return url
    }

    @objc public var activeBundlePath: String? {
        return defaults.string(forKey: Key.active)
    }

    @objc public func setActiveBundlePath(_ path: String) {
        defaults.set(path, forKey: Key.active)
    }

    @objc public func clearActiveBundlePath() {
        defaults.removeObject(forKey: Key.active)
    }

    // ─── Pending bundle ───────────────────────────────────────────────────────

    @objc public var pendingBundlePath: String? {
        return defaults.string(forKey: Key.pending)
    }

    @objc public func setPendingBundlePath(_ path: String) {
        defaults.set(path, forKey: Key.pending)
    }

    @objc public func clearPendingBundlePath() {
        defaults.removeObject(forKey: Key.pending)
    }

    // ─── Previous bundle (rollback) ───────────────────────────────────────────

    @objc public var previousBundlePath: String? {
        return defaults.string(forKey: Key.previous)
    }

    @objc public func setPreviousBundlePath(_ path: String?) {
        if let p = path {
            defaults.set(p, forKey: Key.previous)
        } else {
            defaults.removeObject(forKey: Key.previous)
        }
    }

    // ─── Crash counter ────────────────────────────────────────────────────────

    @objc public var crashCount: Int {
        return defaults.integer(forKey: Key.crashCount)
    }

    @objc public func incrementCrashCount() {
        defaults.set(crashCount + 1, forKey: Key.crashCount)
    }

    @objc public func resetCrashCount() {
        defaults.set(0, forKey: Key.crashCount)
    }

    // ─── Rollback ─────────────────────────────────────────────────────────────

    @objc public func rollback() {
        if let prev = previousBundlePath,
           FileManager.default.fileExists(atPath: prev) {
            setActiveBundlePath(prev)
        } else {
            clearActiveBundlePath()
        }
        clearPendingBundlePath()
        defaults.removeObject(forKey: Key.previous)
    }

    // ─── OTA files directory ──────────────────────────────────────────────────

    /// Returns (creating if needed) <Documents>/ota/<label>/
    @objc public func getOrCreateOtaDir(label: String) throws -> URL {
        let docs = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
        let dir = docs.appendingPathComponent("ota/\(label)", isDirectory: true)
        try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        return dir
    }
}
